### ISSUE TITLE
GdalDecoder::checkSignature has type-o

### ISSUE THREAD

**Issue by shawks03** [opencv/opencv#6567]
Discovered this bug when I built opencv without jpg support but with gdal support. Then trying to open a jpg the gdal decoder ran, and I got an exception from substr. Changing 140 to 0 made the exception go away.

Here's the code within modules/imgcodecs/src/grfmt_gdal.cpp, function GdalDecoder::checkSignature:

``` cpp
bool GdalDecoder::checkSignature( const String& signature )const{
    std::string str = signature.c_str();
    if( str.substr(0,4).find("NITF") != std::string::npos ){
        return true;
    }
    if( str.substr(140,4) == "DTED" ){
        return true;
    }
    return false;
}
```

**Comment by kevin-coder** [comment]
@shawks03 , could you please provide code snip in order to re-produce this issue?

**Comment by shawks03** [comment]
I don't have the original code, but if I remember correctly, I was calling cv::imread on a jpg. And I built my opencv without jpg support, but with gdal support. Therefore, GdalDecoder::checkSignature was called on the jpg, which caused the second substr to throw because the length of signature was only 4. From looking at the code, it's pretty obvious to me that 140 is a mistake.

**Comment by kevin-coder** [comment]
Got you, and seems that it still cannot invoking checkSignature in GdalDecoder, could you please double confirm my re-produce environment as well as steps? Thanks!
**Environment:**
Windows 8.1, CMake 3.5.2 and VS2012
**Reproduce steps:**
1. Clone the latest master branch code and configure in CMake. The key steps in CMake configuration is:
1). Disable the MACRO "WITH_JPEG"
2). Enable the MACRO "WITH_GDAL"
2. After Configure and Generate successfully, build the OpenCV source code.
3. Write sample code which only include calling of function imread, passing the classic "lena.jpg" as its parameter.
4. Debug this sample code and find it cannot calling checkSignature in GdalDecoder.

In this procedure, it called checkSignature functions in other modules exclude GdalDecoder. I think my result was different from yours probably caused by:
1). My reproduce environment is different from yours.
2). My CMake configuration settings might be different from yours.
So, could you please double confirm above 2 steps? Thanks!

**Comment by shawks03** [comment]
Can you confirm that gdal is installed and found by the cmake step for
opencv?

I'm using Ubuntu, but that shouldn't matter.

**Comment by mshabunin** [comment]
I've reproduced the issue. Problem is in this line:
`std::string str = signature.c_str();`
Signature can contain zeroes and std::string constructed from `c_str` has length less than 140 (while GDAL decoder [reserves](https://github.com/Itseez/opencv/blob/master/modules/imgcodecs/src/grfmt_gdal.cpp#L168-L171) 160 bytes for header matching).
The issue can be fixed by using `String` to `string` conversion :smile: : `std::string str = signature;`

I can not say anything about `DTED` signature location, it can possibly start from 140th byte from the beginning. If anyone can find the specs or related code in GDAL - then it should be fixed too.

**Comment by mshabunin** [comment]
Fixed in #7064
